File: app/src/main/java/com/iceteaviet/fastfoodfinder/data/remote/store/model/Store.java
package com.iceteaviet.fastfoodfinder.data.remote.store.model;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.android.gms.maps.model.LatLng;
import com.google.firebase.database.Exclude;
import com.google.firebase.database.PropertyName;
import com.iceteaviet.fastfoodfinder.data.local.store.model.StoreEntity;

/**
 * Created by Genius Doan on 11/10/2016.
 */
public class Store implements Parcelable {

    //Type
    //0: circle_k
    //1: logo_ministop_red
    //2: family mart
    //3: logo_bsmart_red
    //4: shop n go

    public static final Creator<Store> CREATOR = new Creator<Store>() {
        @Override
        public Store createFromParcel(Parcel in) {
            return new Store(in);
        }

        @Override
        public Store[] newArray(int size) {
            return new Store[size];
        }
    };
    @Exclude
    private int type;
    @PropertyName("id")
    private int id;
    @PropertyName("title")
    private String title;
    @PropertyName("address")
    private String address;
    @PropertyName("lat")
    private String lat;
    @PropertyName("lng")
    private String lng;
    @PropertyName("tel")
    private String tel;
    @Exclude
    private LatLng position;

    public Store() {
    }

    public Store(int id, String title, String address, String lat, String lng, String tel, int type) {
        this.id = id;
        this.title = title;
        this.address = address;
        this.lat = lat;
        this.lng = lng;
        this.tel = tel;
        this.type = type;
    }

    public Store(StoreEntity entity) {
        id = entity.getId();
        title = entity.getTitle();
        address = entity.getAddress();
        lat = String.valueOf(entity.getLatitude());
        lng = String.valueOf(entity.getLongitude());
        tel = entity.getTelephone();
        type = entity.getType();
    }


    protected Store(Parcel in) {
        type = in.readInt();
        id = in.readInt();
        title = in.readString();
        address = in.readString();
        lat = in.readString();
        lng = in.readString();
        tel = in.readString();
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    @Exclude
    public LatLng getPosition() {
        if (position == null)
            position = new LatLng(Double.valueOf(lat), Double.valueOf(lng));
        return position;
    }

    public String getTitle() {
        return title;
    }

    @PropertyName("title")
    public void setTitle(String title) {
        this.title = title;
    }

    public String getLat() {
        return lat;
    }

    @PropertyName("lat")
    public void setLat(String lat) {
        this.lat = lat;
    }

    public String getLng() {
        return lng;
    }

    @PropertyName("lng")
    public void setLng(String lng) {
        this.lng = lng;
    }

    @Exclude
    public int getType() {
        return type;
    }

    @Exclude
    public void setType(int type) {
        this.type = type;
    }

    public String getTel() {
        return tel;
    }

    @PropertyName("tel")
    public void setTel(String tel) {
        this.tel = tel;
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Store) {
            return this.id == ((Store) obj).id
                    && this.title.equals(((Store) obj).title);
        }

        return false;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeInt(type);
        parcel.writeInt(id);
        parcel.writeString(title);
        parcel.writeString(address);
        parcel.writeString(lat);
        parcel.writeString(lng);
        parcel.writeString(tel);
    }
}

File: app/src/main/kotlin/com/iceteaviet/fastfoodfinder/data/remote/store/model/Store.kt
package com.iceteaviet.fastfoodfinder.data.remote.store.model

import android.os.Parcel
import android.os.Parcelable
import com.google.android.gms.maps.model.LatLng
import com.google.firebase.database.Exclude
import com.google.firebase.database.PropertyName
import com.iceteaviet.fastfoodfinder.data.local.store.model.StoreEntity

/**
 * Created by Genius Doan on 11/10/2016.
 */
class Store : Parcelable {
    @Exclude
    @get:Exclude
    @set:Exclude
    var type: Int = 0
    @PropertyName("id")
    var id: Int = 0
    @PropertyName("title")
    @set:PropertyName("title")
    var title: String = ""
    @PropertyName("address")
    var address: String = ""
    @PropertyName("lat")
    @set:PropertyName("lat")
    var lat: String = ""
    @PropertyName("lng")
    @set:PropertyName("lng")
    var lng: String = ""
    @PropertyName("tel")
    @set:PropertyName("tel")
    var tel: String = ""
    @Exclude
    private lateinit var position: LatLng

    constructor()

    constructor(id: Int, title: String, address: String, lat: String, lng: String, tel: String, type: Int) {
        try {
            this.id = id
            this.title = title
            this.address = address
            this.lat = lat
            this.lng = lng
            this.tel = tel
            this.type = type

            position = LatLng(lat.toDouble(), lng.toDouble())
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }
    }

    constructor(entity: StoreEntity) : this(entity.id, entity.title, entity.address, entity.latitude.toString(), entity.longitude.toString(), entity.telephone, entity.type)

    protected constructor(`in`: Parcel) : this(`in`.readInt(), `in`.readString()
            ?: "", `in`.readString() ?: "", `in`.readString() ?: "", `in`.readString()
            ?: "", `in`.readString() ?: "", `in`.readInt())


    @Exclude
    fun getPosition(): LatLng {
        if (!::position.isInitialized)
            position = LatLng(lat.toDouble(), lng.toDouble())
        return position
    }

    override fun equals(other: Any?): Boolean {
        return if (other is Store) {
            this.id == other.id && this.title == other.title
        } else false

    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, i: Int) {
        parcel.writeInt(id)
        parcel.writeString(title)
        parcel.writeString(address)
        parcel.writeString(lat)
        parcel.writeString(lng)
        parcel.writeString(tel)
        parcel.writeInt(type)
    }

    override fun hashCode(): Int {
        var result = type
        result = 31 * result + id
        result = 31 * result + title.hashCode()
        result = 31 * result + address.hashCode()
        result = 31 * result + lat.hashCode()
        result = 31 * result + lng.hashCode()
        result = 31 * result + tel.hashCode()
        result = 31 * result + position.hashCode()
        return result
    }

    companion object CREATOR : Parcelable.Creator<Store> {
        // Type
        // 0: circle_k
        // 1: ministop
        // 2: family mart
        // 3: bsmart
        // 4: shop n go

        override fun createFromParcel(parcel: Parcel): Store {
            return Store(parcel)
        }

        override fun newArray(size: Int): Array<Store?> {
            return arrayOfNulls(size)
        }
    }
}
